ink/crates/phat_rollup_anchor_ink/src/traits/meta_transaction.rs

```diff
@@ -0,0 +1,173 @@
+use crate::traits::rollup_anchor::{RollupAnchor, RollupAnchorError, RolupCondEqMethodParams};
+use ink::env::hash::{Blake2x256, HashOutput};
+use ink::prelude::vec::Vec;
+use openbrush::contracts::access_control::{self, AccessControlError, RoleType};
+use openbrush::storage::Mapping;
+use openbrush::traits::{AccountId, Hash, Storage};
+
+pub type Nonce = u128;
+pub type PrepareResult = (ForwardRequest, Hash);
+pub type MetatTxRolupCondEqMethodParams = (ForwardRequest, [u8; 65]);
+
+pub const MANAGER_ROLE: RoleType = ink::selector_id!("MANAGER_ROLE");
+
+#[derive(Debug, Eq, PartialEq, scale::Encode, scale::Decode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub enum MetaTransactionError {
+    NonceTooLow,
+    IncorrectSignature,
+    PublicKeyNotMatch,
+    PublicKeyNotRegistered,
+    AccessControlError(AccessControlError),
+    RollupAnchorError(RollupAnchorError),
+}
+
+/// convertor from AccessControlError to MetaTxError
+impl From<AccessControlError> for MetaTransactionError {
+    fn from(error: AccessControlError) -> Self {
+        MetaTransactionError::AccessControlError(error)
+    }
+}
+
+/// convertor from RollupAnchorError to MetaTxError
+impl From<RollupAnchorError> for MetaTransactionError {
+    fn from(error: RollupAnchorError) -> Self {
+        MetaTransactionError::RollupAnchorError(error)
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, scale::Encode, scale::Decode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub struct ForwardRequest {
+    pub from: AccountId,
+    pub nonce: Nonce,
+    pub data: Vec<u8>,
+}
+
+#[derive(Default, Debug)]
+#[openbrush::storage_item]
+pub struct Data {
+    nonces: Mapping<AccountId, Nonce>,
+    ecdsa_public_keys: Mapping<AccountId, Vec<u8>>,
+}
+
+#[openbrush::trait_definition]
+pub trait MetaTransaction:
+    Storage<Data> + EventBroadcaster + access_control::Internal + RollupAnchor
+{
+    #[ink(message)]
+    fn get_ecdsa_public_key(&self, from: AccountId) -> [u8; 33] {
+        self.data::<Data>()
+            .ecdsa_public_keys
+            .get(&from)
+            .map(|k| k.try_into().unwrap_or([0; 33]))
+            .unwrap_or([0; 33])
+    }
+
+    #[ink(message)]
+    #[openbrush::modifiers(access_control::only_role(MANAGER_ROLE))]
+    fn register_ecdsa_public_key(
```

Comment:
Not necessary. You need to register the public key, because you don't know how the account `from` and the `ecdsa_public_key` are mapped. However, usually the `from` account id is calculated from the public key. So you can just verify the signature is done by the signer (attestor) account id. No need to register the public key again.

Reply:
It's expected soon: https://github.com/paritytech/ink/pull/1840

Reply:
Yes, using a sr25519 key is the easiest way. However even with seco256k1, it's still possible to calculate an AccountId based on its public key.

For example, following Substrate's ecdsa format, the address is the Blake2 hash of the compressed 33 bytes public key. So if you submit the message and the signature, you can recover the public key, and then calculate the AccountId. The AccountId can be added to the ATTESTOR role for access control.

Reply:
I remember trying to do something similar, but it failed. I will try again to understand my mistake. 
Can we consider that the current implementation is sufficient to pass milestone 1 knowing that I would use the sr25519 signature verification planned in the next version of Ink!.

Reply:
Yes, totally fine.